community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/Pattern.scala

```diff
@@ -274,12 +274,27 @@ case class RelationshipPattern(
     }
 
   def semanticCheck(ctx: SemanticContext): SemanticCheck =
+    checkNoInvalidSelfReferences chain
     checkNoOptionalRelsForAnExpression(ctx) chain
-    checkNoVarLengthWhenUpdating(ctx) chain
-    checkNoLegacyOptionals(ctx) chain
-    checkNoParamMapsWhenMatching(ctx) chain
-    checkProperties(ctx) chain
-    checkNotUndirectedWhenCreating(ctx)
+      checkNoVarLengthWhenUpdating(ctx) chain
+      checkNoLegacyOptionals(ctx) chain
+      checkNoParamMapsWhenMatching(ctx) chain
+      checkProperties(ctx) chain
+      checkNotUndirectedWhenCreating(ctx)
+
+
+  private def checkNoInvalidSelfReferences: SemanticCheck = {
+    when(!isSingleLength) {
+      val (hasSelfReference, pos) = properties.treeFold((false, position)) {
```

Comment:
I am afraid that using a `treeFold` this was you are gonna report a semantic error for this case as well: `MATCH a, b, (a)-[r* {r: '42'}]->(b)`, which should work just fine. Perhaps we should add that test and make sure it does not fail.